modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/delegate/delegate.es.js

```diff
@@ -0,0 +1,59 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+/**
+ * Checks if element is disabled or whether it exists in a disabled element tree
+ * @param {!Element} element The DOM element to check.
+ * @return {boolean}
+ */
+function isDisabled(node) {
+	return !!(node.disabled || node.closest('[disabled]'));
+}
+
+/**
+ * Listens to the specified event on the given DOM element, but only calls the
+ * given callback listener when it's triggered by elements that match the
+ * given selector or target element.
+ * @param {!Element} element The DOM element the event should be listened on.
+ * @param {string} eventName The name of the event to listen to.
+ * @param {string} selector Css selector that should match the event for the
+ *     listener to be triggered.
+ * @param {!function(!Object)} callback Function to be called when the event
+ *     is triggered. It will receive the normalized event object.
+ * @return {function} Can be used to remove the listener.
+ */
+function delegate(element, eventName, selector, callback) {
+	const eventHandler = (event) => {
+		if (eventName === 'click' && isDisabled(event.target)) {
+			return;
+		}
+
+		if (event.target.matches(selector)) {
+			callback(event);
+		}
+	};
+
+	element.addEventListener(eventName, eventHandler);
+
+	const removeEvent = () => {
+		element.removeEventListener(eventName, eventHandler);
+	};
+
+	removeEvent.dispose = removeEvent;
```

Comment:
Doesn't this feel a bit weird? 😂 

Not blocking, but not sure if we do this elsewhere. What's the reason to add a property to an arrow function pointing to itself? 

Is it so that it can be used in a dual fashion? Do we want this to be used in a dual fashion?

Reply:
I don't have a strong opinion on this, but I would probably prefer two things:

1.  Use the same name; ie. don't have the same thing named `removeEvent` in one scope and `dispose` somewhere else.
2.  But more than that, I'd prefer that we had one way to do things. In this context, it seems like returning the function is to make it easier to use in a React `useEffect()` hook. So, if you want to do that, I would do it one way here:
    
    ```js
    return {
        dispose() {
            element.removeEventListener(eventName, eventHandler);
        },
    };
    ```
    
    so that callers can do:
    
    ```js
    // this...
    const {dispose} = delegate(...);
    
    // so that they can later do...
    dispose();
    ```
    
    (only returning an object with a `dispose` method on it because it provides us a little bit more flexibility to evolve the API in the future, should we need to).
    
    And then in `frontend-js-react-web` I'd offer a separate `useDelegate()` hook that people could use instead, and which would abstract over the need to worry about `useEffect()` anyway; eg.
    
    ```js
    function useDelegate(...) {
        useEffect(() => {
            const {dispose} = delegate(...);
            
            return dispose;
        };
    }
    ```